lib/cdo/i18n_backend.rb

```diff
@@ -164,12 +164,118 @@ def translate(locale, key, options = ::I18n::EMPTY_HASH)
+      class ::I18n::Backend::LocaleExtractor
+        def self.locale_from_path(path)
+          # Extracts the locale from the path, like "en", "en-US", "haw", "haw-HI", etc.
+          path[/\b([a-z]{2,3}(?:-[A-Z]{2})?)\b\.\w+$/, 1]&.to_sym
+        end
+      end
+
+      def load_translations(*files)
+        loaded_files.merge(files.flatten) if lazy_load?
+        super
+      end
+
+      def reload!
+        if lazy_load?
+          @loaded_files = nil
+          @valid_locales = nil
+        end
+
+        super
+      end
+
+      def eager_load!
+        # Ignores eager loading instead of raising an error if lazy loading is enabled
+        super unless lazy_load?
+      end
+
+      protected
+
+      def loaded_files
+        @loaded_files ||= Set.new
+      end
+
+      # Get the set of valid locales for a given base locale code. For example,
+      # for a base of `:pt`, we expect valid locales `:pt`, `:"pt-BR"`, `:"en-US"`, and `en`
+      def valid_locales_for(base_locale)
+        @valid_locales ||= {}
+
+        unless @valid_locales[base_locale]
+          @valid_locales[base_locale] = Set.new(::I18n.fallbacks[base_locale])
+          @valid_locales[base_locale] << base_locale
+          @valid_locales[base_locale] << LOCALES_MAPPING[base_locale] # en: :'en-US'
+          @valid_locales[base_locale] = @valid_locales[base_locale].compact_blank
+        end
+
+        @valid_locales[base_locale]
+      end
+
+      # The original method has been modified to also load files for the current locale's fallbacks,
+      # like loading "en-US" files for the "en" locale and vice versa.
+      # https://github.com/ruby-i18n/i18n/blob/v1.12.0/lib/i18n/backend/lazy_loadable.rb#L164-L171
+      def filenames_for_current_locale
+        valid_locales = valid_locales_for(::I18n.locale)
+        # Excludes already loaded i18n files to prevent them from being reloaded during locale switching.
+        # For example, for the "de-DE" locale, the i18n fallbacks are "de", "en", and "en-US".
+        # Since "en-US" (and "en") is the default locale, its i18n files are already loaded.
+        (::I18n.load_path.flatten - loaded_files.to_a).select do |path|
+          path_locale = ::I18n::Backend::LocaleExtractor.locale_from_path(path)
+          path_locale.nil? || valid_locales.include?(path_locale)
+        end
+      end
+
+      # Checks if a filename corresponds to the locale it has loaded.
+      # The locale extracted from the path must be either the single locale loaded in
+      # the translations or one of the expected variations of the locale, taking into
+      # account both long locale codes and fallbacks.
+      #
+      # The original implementation was changed from raising an error to log the details to the console,
+      # because the error isn't all that critical to catch and
+      # the new loading logic includes more weird edge cases.
```

Comment:
nit: grammar and line wrapping
```suggestion
      # The original implementation was changed from raising an error to logging
      # the details to the console, because the error isn't all that critical
      # to catch and the new loading logic includes more weird edge cases.
```